pkg/kv/kvserver/gc/gc.go

```diff
@@ -354,7 +366,26 @@ func processReplicatedKeyRange(
 	options intentBatcherOptions,
 	cleanupIntentsFn CleanupIntentsFunc,
 	info *Info,
-) error {
+) (bool, error) {
+	// Perform fast path check prior to performing GC. Fast path only collects
+	// user key span portion, so we don't need to clean it up once again if
+	// we succeeded.
+	collectUserKeySpan := true
+	{
+		start := desc.StartKey.AsRawKey()
+		end := desc.EndKey.AsRawKey()
+		if err := storage.CheckRangeEmptiness(ctx, snap, start, end, threshold); err == nil {
+			if err = gcer.GC(ctx, nil, nil, &roachpb.GCRequest_GCClearRangeKey{
+				StartKey: start,
+				EndKey:   end,
+			}); err == nil {
+				collectUserKeySpan = false
+				info.FastPathOperations++
+				return true, nil
```

Comment:
If we're returning here, won't we bypass GC of local data?

Reply:
I assume so, otherwise I don't understand why we bother with iterating over them here.

Reply:
Range descriptors, for example, most certainly use transactions and thus MVCC, and will need GC.